IDGPredict/FacetPredict.h

```diff
@@ -157,51 +172,83 @@ class FacetPredict
+    size_t nTerms = _readers.size();
+    typedef std::vector<std::pair<size_t, std::complex< float >*>> rowidlist_t;
+    std::vector<rowidlist_t> available_row_ids(nTerms);
+    for(size_t term=0; term!=nTerms; ++term)
+    {
+      idg::api::BufferSet& bs = *_buffersets[direction*nTerms + term];
+      available_row_ids[term] = bs.get_degridder(dataDescId)->compute();
+    }
+    
+    size_t nChan = _bands[dataDescId].size();
+    double
+      dlFact = 2.0 * M_PI * _metaData[direction].dl,
+      dmFact = 2.0 * M_PI * _metaData[direction].dm,
+      dpFact = 2.0 * M_PI * _metaData[direction].dp;
+    for(size_t i=0; i!=available_row_ids[0].size(); ++i)
     {
-      size_t row = i.first;
-      std::complex<float>* values = i.second;
-      size_t nChan = _bands[dataDescId].size();
+      size_t row = available_row_ids[0][i].first;
       size_t localRow = row - _metaData[direction].rowIdOffset;
-      double* uvw = &_metaData[direction].uvws[localRow*3];
-      double
-        dlFact = 2.0 * M_PI * _metaData[direction].dl,
-        dmFact = 2.0 * M_PI * _metaData[direction].dm,
-        dpFact = 2.0 * M_PI * _metaData[direction].dp;
+      const double* uvw = &_metaData[direction].uvws[localRow*3];
+      
+      // Correct the phase shift of the values for this facet
+      for(size_t term=0; term!=nTerms; ++term)
+      {
+        std::complex<float>* values = available_row_ids[term][i].second;
+        for(size_t ch=0; ch!=nChan; ++ch)
+        {
+          double
+            lambda = wavelength(dataDescId, ch),
+            u = uvw[0] / lambda,
+            v = uvw[1] / lambda,
+            w = uvw[2] / lambda;
+          double angle = u * dlFact + v * dmFact + w * dpFact;
+          float
+            rotSin = sin(angle),
+            rotCos = cos(angle);
+          for(size_t p=0; p!=4; ++p) {
+            std::complex<float> s = values[ch*4 + p];
+            values[ch*4 + p] = std::complex<float>(
+              s.real() * rotCos  -  s.imag() * rotSin,
+              s.real() * rotSin  +  s.imag() * rotCos);
+          }
+        }
+      }
+      
+      // Apply polynomial-term corrections and add all to values of 'term 0'
+      std::complex<float>* values0 = available_row_ids[0][i].second;
       for(size_t ch=0; ch!=nChan; ++ch)
       {
-        double
-          lambda = wavelength(dataDescId, ch),
-          u = uvw[0] / lambda,
-          v = uvw[1] / lambda,
-          w = uvw[2] / lambda;
-        double angle = u * dlFact + v * dmFact + w * dpFact;
-        float
-          rotSin = sin(angle),
-          rotCos = cos(angle);
-        for(size_t p=0; p!=4; ++p) {
-          std::complex<float> s = values[ch*4 + p];
-          values[ch*4 + p] = std::complex<float>(
-            s.real() * rotCos  -  s.imag() * rotSin,
-            s.real() * rotSin  +  s.imag() * rotCos);
+        double frequency = _bands[dataDescId][ch];
+        double freqFactor = frequency / _refFrequency - 1.0;
```

Comment:
Is this the definition of spectral index used? If it is not very standard, it would be good to add this to either documentation or comments.

Reply:
I've added:

      // The "polynomial spectrum" definition is used, equal to the one e.g. used by WSClean in component outputs
      // (see https://sourceforge.net/p/wsclean/wiki/ComponentList/ ) and in text files when 'logarithmic SI' is false. 
      // The definition is:
      //   S(nu) = term0 + term1 (nu/refnu - 1) + term2 (nu/refnu - 1)^2 + ...